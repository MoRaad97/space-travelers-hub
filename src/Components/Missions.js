import React from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { missionsReducer } from '../Redux/Missions/Missions';
import './style/missions.css';

const Missions = () => {
  // get rockets data from the store
  const missions = useSelector((state) => state.mission);

  const dispatch = useDispatch();

  const handle = (e) => {
    const { id } = e.target;
    dispatch(missionsReducer.joinAction({ id }));
  };

  return (
    <section className="missions">
      <div className="container">
        <table className="table-bordered">
          <thead>
            <tr>
              <td>Mission</td>
              <td>Description</td>
              <td>Status</td>
              <td> </td>
            </tr>
          </thead>
          <tbody>
            {missions.map((mission) => (
              <tr className="mission-card" key={mission.mission_id}>
                <td>
                  <p className="mission-name">{mission.mission_name}</p>
                </td>
                <td>
                  {mission.description}
                </td>
                <td>
                  {!mission.joined && <p className="not-member" id="member" type="button"> NOT A MEMBER</p>}
                  {mission.joined && <p className="active-member" id="member" type="button"> ACTIVE MEMBER</p>}
                </td>
                <td>
                  {!mission.joined && (
                    <button id={mission.mission_id} type="button" className="not-joined" onClick={handle}>
                      JOIN MISSION
                    </button>
                  )}
                  {mission.joined && (
                    <button id={mission.mission_id} type="button" className="joined" onClick={handle}>
                      LEAVE MISSION
                    </button>
                  )}
                </td>

              </tr>
            ))}
          </tbody>
        </table>
      </div>
    </section>

  );
};

export default Missions;
