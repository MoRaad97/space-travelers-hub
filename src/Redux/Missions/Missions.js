import axios from 'axios';
import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';

// Actions
const LOAD_MISSIONS = 'spacex-project-solo/rockets/LOAD_MISSIONS';

// Url
const baseUrl = 'https://api.spacexdata.com/v3/missions';

// create ASYNC Thunks
export const loadMissionsThunk = createAsyncThunk(LOAD_MISSIONS, async () => {
  // fetch Misisons
  const response = await axios.get(`${baseUrl}`).catch((err) => {
    console.log('Error', err);
  });
  const res = response.data;
  return res;
});

const missionsSlice = createSlice({
  name: 'spacex-project-solo/missions',
  initialState: [],
  reducers: {
    joinAction(state, action) {
      const newState = state.map((object) => {
        if (object.mission_id === action.payload.id) {
          return {
            ...object,
            joined: !object.joined,
          };
        }
        return {
          ...object,
        };
      });
      return newState;
    },
  },
  extraReducers: {
    [loadMissionsThunk.fulfilled]: (state, action) => {
      const newState = action.payload.map((object) => ({
        mission_id: object.mission_id,
        mission_name: object.mission_name,
        description: object.description,
        joined: false,
      }));
      return newState;
    },
  },
});

export const missionsReducer = missionsSlice.actions;
export default missionsSlice.reducer;
