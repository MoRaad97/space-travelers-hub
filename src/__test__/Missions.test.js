import React from 'react';
import TestRenderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import store from '../Redux/configureStore';
import Missions from '../Components/Missions';

it('Missions Component snapshot test', () => {
  const missionsRender = TestRenderer
    .create(
      <Provider store={store}>
        <Missions />
      </Provider>,
    )
    .toJSON();
  expect(missionsRender).toMatchSnapshot();
});
