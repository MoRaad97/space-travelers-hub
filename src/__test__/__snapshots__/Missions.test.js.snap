// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Missions Component snapshot test 1`] = `
<section
  className="missions"
>
  <div
    className="container"
  >
    <table
      className="table-bordered"
    >
      <thead>
        <tr>
          <td>
            Mission
          </td>
          <td>
            Description
          </td>
          <td>
            Status
          </td>
          <td>
             
          </td>
        </tr>
      </thead>
      <tbody />
    </table>
  </div>
</section>
`;
